### PR TITLE
Data Dictionary: Migration for group field to a model

### PR DESCRIPTION
**Pull request by pxwxnvermx**
## Technical Summary
This PR adds migration for creating the CasePropertyGroup model, foreign key for group_id on CaseProperty model and a management command to populate group data in CasePropertyGroup model from CaseProperty model. The data migration for populating group data to model and foreign keys would be run async using the command in this PR.

[Ticket Link](https://dimagi-dev.atlassian.net/browse/USH-2979)

## Feature Flag
[DATA_DICTIONARY](https://staging.commcarehq.org/hq/flags/edit/data_dictionary/)

### Safety story

These changes have been tested locally and on staging.
### Migrations

- [x] The migrations in this code can be safely applied first independently of the code

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

File: corehq/apps/data_dictionary/management/commands/populate_case_prop_groups.py
Comment (mkangia) [inline corehq/apps/data_dictionary/management/commands/populate_case_prop_groups.py:25]: should this be group_id__exact?

File: corehq/apps/data_dictionary/migrations/0012_caseproperty_group_id.py
Comment (mkangia) [inline corehq/apps/data_dictionary/migrations/0012_caseproperty_group_id.py:1]: nit: You could merge this with the previous commit.

File: corehq/apps/data_dictionary/migrations/0011_casepropertygroup.py
Comment (mkangia) [inline corehq/apps/data_dictionary/migrations/0011_casepropertygroup.py:1]: you could merge this with the commit that added the migrations originally.

**Review by mkangia — COMMENTED** [review]
Looks good with first pass. Just some minor queries/feedback

Just something to flag, in terms of efficiency, we are calling `get_or_create` for CaseType for each case property group to be created. Can they be grouped by case type so that we don't call it again and again, or not call it at all, since CaseType would always be present anyway during the migration.

File: corehq/apps/data_dictionary/models.py
Comment (mkangia) [inline corehq/apps/data_dictionary/models.py:46]: why is this related_name "groups"?

File: corehq/apps/data_dictionary/models.py
Comment (mkangia) [inline corehq/apps/data_dictionary/models.py:103]: why is this related name "properties"?

File: corehq/apps/data_dictionary/management/commands/populate_case_prop_groups.py
Comment (pxwxnvermx) [reply corehq/apps/data_dictionary/management/commands/populate_case_prop_groups.py:25]: the `group` field is blank i.e empty string for all the case properties that don't have a group, `group__exact=""` filters them. I can add one more statement for filtering all the fields that were already assigned a group_id this will make this command idempotent.

File: corehq/apps/data_dictionary/management/commands/populate_case_prop_groups.py
Comment (pxwxnvermx) [reply corehq/apps/data_dictionary/management/commands/populate_case_prop_groups.py]: Point 2. seems to be a good solution. We can also filter out the properties with null group_id in the query to database itself, that will make this command idempotent and on resumes only the fields that are left.

File: corehq/apps/data_dictionary/management/commands/populate_case_prop_groups.py
Comment (mkangia) [reply corehq/apps/data_dictionary/management/commands/populate_case_prop_groups.py:25]: Is there a difference in how the query is run for `group_id__exact` and `group__exact`?

File: corehq/apps/data_dictionary/models.py
Comment (millerdev) [inline corehq/apps/data_dictionary/models.py:100]: I think this should be named `group` (Django automatically adds the `_id` suffix on the db column). Then `db_column="group_id"` can be removed below, and the related group object can be referenced by `case_property.group`.

File: corehq/apps/data_dictionary/migrations/0012_caseproperty_group_id.py
Comment (millerdev) [reply corehq/apps/data_dictionary/migrations/0012_caseproperty_group_id.py:1]: Did you mean squash this into the previous migration? I think that's what should be done here. There will be more changes to be incorporated into the migration as well if `group_id` is renamed to `group`.

The easiest way to do that is to un-apply these migrations on any database where they have been applied:
```sh
# unapply both migrations from this PR
# WARNING this is a destructive command. The `CasePropertyGroup` table will be dropped.
./manage.py migrate data_dictionary 0010
```
Then on your development workstation, remove the two migration files
```sh
rm corehq/apps/data_dictionary/migrations/0012_caseproperty_group_id.py
rm corehq/apps/data_dictionary/migrations/0011_casepropertygroup.py
```
And remove the relevant lines from the `migrations.lock` file.

Finally, re-generate a single new migration
```
./mange.py makemigrations data_dictionary -n casepropertygroup
```

File: corehq/apps/data_dictionary/models.py
Comment (millerdev) [reply corehq/apps/data_dictionary/models.py:60]: I agree. I think it is unnecessary to use `CriticalSection`; `get_or_create` should do what you want.

File: corehq/apps/data_dictionary/migrations/0012_caseproperty_group_id.py
Comment (pxwxnvermx) [reply corehq/apps/data_dictionary/migrations/0012_caseproperty_group_id.py:1]: Thanks @millerdev @mkangia  for the inputs. I will squash the migrations and add another commit for it.

File: corehq/apps/data_dictionary/management/commands/populate_case_prop_groups.py
Comment (pxwxnvermx) [reply corehq/apps/data_dictionary/management/commands/populate_case_prop_groups.py:25]: we can use `group_id__isnull=True` in the `filter` method to filter out all the records which have a null group_id.

File: corehq/apps/data_dictionary/models.py
Comment (millerdev) [reply corehq/apps/data_dictionary/models.py:100]: Ah, I missed the `group` text column. However, I think this name will be confusing because  `case_property.group_id` will return a `CasePropertyGroup` object (not an id, which is typically an integer or a string).

File: corehq/apps/data_dictionary/models.py
Comment (pxwxnvermx) [reply corehq/apps/data_dictionary/models.py:60]: Offlined with @mkangia, I don't think that custom `get_or_create` would be required for this because django already provides this method.

File: corehq/apps/data_dictionary/models.py
Comment (pxwxnvermx) [reply corehq/apps/data_dictionary/models.py:64]: Yeah, it would be present before the group. Thanks for pointing out.

File: corehq/apps/data_dictionary/models.py
Comment (pxwxnvermx) [reply corehq/apps/data_dictionary/models.py:100]: Yeah, the name can be a little confusing, i can make a comment regarding what `group_id` returns, that can help in reducing this confusion.

fyi @mkangia

File: corehq/apps/data_dictionary/migrations/0012_caseproperty_group_id.py
Comment (pxwxnvermx) [reply corehq/apps/data_dictionary/migrations/0012_caseproperty_group_id.py:1]: https://github.com/dimagi/commcare-hq/pull/32799/commits/9ae5ddf1a7ecfde55be53726774d140a31e6200a

File: corehq/apps/data_dictionary/management/commands/populate_case_prop_groups.py
Comment (pxwxnvermx) [reply corehq/apps/data_dictionary/management/commands/populate_case_prop_groups.py:25]: [f3a51e3](https://github.com/dimagi/commcare-hq/pull/32799/commits/f3a51e32da7363f6d7a48356b39b60f7063515f6)

File: corehq/apps/data_dictionary/models.py
Comment (pxwxnvermx) [reply corehq/apps/data_dictionary/models.py:60]: This change removes the usage of my custom get_or_create method. Also this avoids fetching case_type since it is already present on case property. [c4f4de2](https://github.com/dimagi/commcare-hq/pull/32799/commits/c4f4de2a5cd107415a920ea3f14260830163798a)

File: corehq/apps/data_dictionary/models.py
Comment (mkangia) [reply corehq/apps/data_dictionary/models.py:100]: Hey @millerdev 

@pxwxnvermx and I had some back and forth on the approach to be taken here.

We were initially considering renaming of the old column and adding the new one like you suggested, but that seemed brittle considering that it could lead to issues on the environment in between the time migrations are run and the code update is done for webworkers.
We could also, do the renaming first in a deploy separately but doing any sort of renaming could cause issues because of the reason shared above.
So we decided to go with this approach above.

We would remove the "group" column eventually.
I believe then, we could just rename this `group_id` to group in code in model, which would need no migration since the column name would remain the same (I believe @pxwxnvermx confirmed this locally), and also update any code references from group_id to group on objects.

Open to suggestions/recommendations here.

File: corehq/apps/data_dictionary/management/commands/populate_case_prop_groups.py
Comment (mkangia) [inline corehq/apps/data_dictionary/management/commands/populate_case_prop_groups.py:31]: is this the intended change for this commit?
" Refactor to use CasePropertyGroup.objects.get_or_create "

Comment (pxwxnvermx) [reply corehq/apps/data_dictionary/management/commands/populate_case_prop_groups.py:31]: Yeah, i have used the django inbuilt get_or_create method, i will remove the `get_or_create` method that i created as it is not needed.

Comment (mkangia) [reply corehq/apps/data_dictionary/management/commands/populate_case_prop_groups.py:31]: Ah, right. I missed that "objects" was now added.

File: corehq/apps/data_dictionary/models.py
Comment (mkangia) [inline corehq/apps/data_dictionary/models.py:1]: I think good to squash this with the previous commit so its clear why it was okay to remove

File: corehq/apps/data_dictionary/models.py
Comment (millerdev) [reply corehq/apps/data_dictionary/models.py:85]: I think it's fine to use `group_id` as a transitional name as long as it gets renamed to `group` before this effort is finished. Maybe add a comment to indicate the reason and plan for eventual state?

Comment (millerdev) [reply corehq/apps/data_dictionary/models.py:85]: Thinking about this more, maybe consider using `group_obj` rather than `group_id` to convey that the property returns an object, not an id?

Comment (mkangia) [reply corehq/apps/data_dictionary/models.py:85]: Sounds good to me, I leave the final call to @pxwxnvermx

Comment (pxwxnvermx) [reply corehq/apps/data_dictionary/models.py:85]: Yeah, this looks good to me too.

Comment (pxwxnvermx) [reply corehq/apps/data_dictionary/models.py]: Fixed. [8aa1d2e](https://github.com/dimagi/commcare-hq/pull/32799/commits/8aa1d2e1ebff9ab9f462e49e62166fd5e64d75d9)

File: corehq/apps/data_dictionary/management/commands/populate_case_prop_groups.py
Comment (esoergel) [inline corehq/apps/data_dictionary/management/commands/populate_case_prop_groups.py:24]: nice, this `group_id__isnull` should make this migration idempotent

File: corehq/apps/data_dictionary/models.py
Comment (esoergel) [reply corehq/apps/data_dictionary/models.py:90]: Or does this make it easier to eventually migrate the `ForeignKey` to be just called `group`, since that'd result in a `group_id` column in the DB to store the related model's primary key?

Comment (pxwxnvermx) [reply corehq/apps/data_dictionary/models.py:90]: This makes it easier to rename the `group_obj` field to `group`.

**Review by esoergel — APPROVED** [review]

File: corehq/apps/data_dictionary/management/commands/populate_case_prop_groups.py
Comment (mkangia) [inline corehq/apps/data_dictionary/management/commands/populate_case_prop_groups.py:39]: nit: properties "that" have

File: corehq/apps/data_dictionary/management/commands/populate_case_prop_groups.py
Comment (mkangia) [inline corehq/apps/data_dictionary/management/commands/populate_case_prop_groups.py:44]: Should not this delete the group_obj as well? or they would get removed in the next query?

File: corehq/apps/data_dictionary/management/commands/populate_case_prop_groups.py
Comment (mkangia) [inline corehq/apps/data_dictionary/management/commands/populate_case_prop_groups.py:43]: Good to also show the name of these case properties or some sort of representation so that user can actually check if things are okay. 
Same for the logging later as well

File: corehq/apps/data_dictionary/management/commands/populate_case_prop_groups.py
Comment (pxwxnvermx) [reply corehq/apps/data_dictionary/management/commands/populate_case_prop_groups.py:44]: The next query removes the groups with no properties, we can not delete the group here. Group might have other properties that reference it.

File: corehq/apps/data_dictionary/management/commands/populate_case_prop_groups.py
Comment (pxwxnvermx) [reply corehq/apps/data_dictionary/management/commands/populate_case_prop_groups.py]: Fixed. [e48381a](https://github.com/dimagi/commcare-hq/pull/32799/commits/e48381ad475009f6d6e85d8d8aafdcb67b530ad7)

File: corehq/apps/data_dictionary/management/commands/populate_case_prop_groups.py
Comment (pxwxnvermx) [reply corehq/apps/data_dictionary/management/commands/populate_case_prop_groups.py]: Fixed. [c41883e](https://github.com/dimagi/commcare-hq/pull/32799/commits/c41883e62716997754992920b8be35c42ce7b62d)

File: corehq/apps/data_dictionary/management/commands/populate_case_prop_groups.py
Comment (esoergel) [inline corehq/apps/data_dictionary/management/commands/populate_case_prop_groups.py:17]: What's the story here?  An issue in the sync code?

File: corehq/apps/data_dictionary/management/commands/populate_case_prop_groups.py
Comment (esoergel) [inline corehq/apps/data_dictionary/management/commands/populate_case_prop_groups.py:45]: It'd be helpful to also report domain and case type here if the idea is that these might need following up on

**Review by esoergel — APPROVED** [review]

File: corehq/apps/data_dictionary/management/commands/populate_case_prop_groups.py
Comment (pxwxnvermx) [reply corehq/apps/data_dictionary/management/commands/populate_case_prop_groups.py:17]: This code was added to fix the case where the user changes the group for a property after this merging and migrating this PR but before the Next PR for this change is merged. Here the group field for a property would be out of sync with group_obj reference, so this code resets the group_obj field to None. Then the populate_case_prop_groups function would sync it again.

File: corehq/apps/data_dictionary/management/commands/populate_case_prop_groups.py
Comment (pxwxnvermx) [reply corehq/apps/data_dictionary/management/commands/populate_case_prop_groups.py]: I have added them to the log for these also. [0bc6705](https://github.com/dimagi/commcare-hq/pull/32799/commits/0bc6705fba7078b61eedce72c7fdff76cf8d8109)

**Review by esoergel — APPROVED** [review]

**Pull request merged**
